File: loong's code/leetcode/editor/cn/1604. 警告一小时内使用相同员工卡大于等于三次的人.go
package main

import (
	"fmt"
	"sort"
)

func alertNames(keyName []string, keyTime []string) (ans []string) {
	d := map[string][]int{}
	for i, name := range keyName {
		var a, b int
		// Sscanf从字符串str扫描文本，根据format 参数指定的格式将成功读取的空白分隔的值保存进成功传递给本函数的参数。返回成功扫描的条目个数和遇到的任何错误。
		_, err := fmt.Sscanf(keyTime[i], "%d:%d", &a, &b)
		if err != nil {
			return nil
		}
		t := a*60 + b
		d[name] = append(d[name], t)
	}
	for name, ts := range d {
		n := len(ts)
		if n > 2 {
			sort.Ints(ts)
			for i := 0; i < n-2; i++ {
				if ts[i+2]-ts[i] <= 60 {
					ans = append(ans, name)
					break
				}
			}
		}

	}
	sort.Strings(ans)
	return
}

File: loong's code/leetcode/editor/cn/1604警告一小时内使用相同员工卡大于等于三次的人.py
from collections import defaultdict
from typing import List


class Solution:
    def alertNames(self, keyName: List[str], keyTime: List[str]) -> List[str]:
        ans = set()
        dict1 = defaultdict(list)
        for name, time in zip(keyName, keyTime):
            dict1[name].append(int(time[:2])*60+int(time[3:]))
        print(dict1)
        for i in dict1:
            # 定义10：00-11：00出现三次计数，就警告
            if len(dict1[i]) >= 3:
                dict1[i].sort()
                print(dict1[i])
                for j in range(len(dict1[i])):
                    # 遍历当前小时到一小时内,好像可以直接看切片
                    # for k in range(len(dict1[i]) + 1, len(dict1[i])+3):
                    if j + 2 < len(dict1[i]):
                        # if dict1[i][j:j + 3][0].replace(":", "")[:2]
                        if abs(dict1[i][j:j + 3][-1] -
                            dict1[i][j:j + 3][0]) <= 60:
                            ans.add(i)
                        print(dict1[i][j:j + 3])
        ans = list(ans)
        ans.sort()
        return ans


print(Solution.alertNames(Solution(), keyName = ["a","a","a","a","a","b","b","b","b","b","b"]
                          , keyTime = ["04:48","23:53","06:36","07:45","12:16","00:52","10:59","17:16","00:36","01:26","22:42"]))
